app/src/main/java/net/squanchy/service/firebase/FirebaseDbService.kt

```diff
@@ -0,0 +1,154 @@
+import net.squanchy.service.firebase.model.FirebaseEvents
+import net.squanchy.service.firebase.model.FirebaseFavorites
+import net.squanchy.service.firebase.model.FirebasePlaces
+import net.squanchy.service.firebase.model.FirebaseSpeakers
+import net.squanchy.service.firebase.model.FirebaseTrack
+import net.squanchy.service.firebase.model.FirebaseTracks
+import net.squanchy.service.firebase.model.FirebaseUserData
+import net.squanchy.service.firebase.model.FirebaseVenue
+
+class FirebaseDbService(private val database: DatabaseReference) {
+
+    fun days(): Observable<FirebaseDays> {
+        return observeChild(daysNode(), FirebaseDays::class.java)
+    }
+
+    fun speakers(): Observable<FirebaseSpeakers> {
+        return observeChild(speakersNode(), FirebaseSpeakers::class.java)
+    }
+
+    fun events(): Observable<FirebaseEvents> {
+        return observeChild(eventsNode(), FirebaseEvents::class.java)
+    }
+
+    fun event(eventId: String): Observable<FirebaseEvent> {
+        return observeChild(eventByIdNode(eventId), FirebaseEvent::class.java)
+    }
+
+    fun places(): Observable<FirebasePlaces> {
+        return observeChild(placesNode(), FirebasePlaces::class.java)
+    }
+
+    fun tracks(): Observable<FirebaseTracks> {
+        return observeChild(tracksNode(), FirebaseTracks::class.java)
+    }
+
+    fun track(trackId: String): Observable<FirebaseTrack> {
+        return observeChild(trackByIdNode(trackId), FirebaseTrack::class.java)
+    }
+
+    fun favorites(userId: String): Observable<FirebaseFavorites> {
+        return userData(userId)
+                .map { (favorites) -> favorites ?: emptyMap() }
+                .map(::FirebaseFavorites)
+    }
+
+    fun achievements(userId: String): Observable<FirebaseAchievements> {
+        return userData(userId)
+                .map { (_, achievements) -> achievements ?: emptyMap() }
+                .map(::FirebaseAchievements)
+    }
+
+    private fun userData(userId: String): Observable<FirebaseUserData> {
+        return observeOptionalChild(userDataNode(userId), FirebaseUserData::class.java, lazy { FirebaseUserData() })
+    }
+
+    fun venueInfo(): Observable<FirebaseVenue> {
+        return observeChild(venueInfoNode(), FirebaseVenue::class.java)
+    }
+
+    private fun <T> observeChild(path: String, clazz: Class<T>): Observable<T> {
+        return observeChildAndEmit(path, clazz, { it!! })
+    }
+
+    private fun <T> observeOptionalChild(path: String, clazz: Class<T>, default: Lazy<T>): Observable<T> {
+        return observeChildAndEmit(path, clazz, { it ?: default.value })
+    }
+
+    private fun <T, V> observeChildAndEmit(path: String, clazz: Class<V>, map: (V?) -> T): Observable<T> {
+        return Observable.create { emitter: ObservableEmitter<T> ->
+            val listener = object : ValueEventListener {
+                override fun onDataChange(dataSnapshot: DataSnapshot) {
+                    if (emitter.isDisposed) {
```

Comment:
They check, but then they will throw an exception if something is sent to a disposed observable. This is a crash we kept having in our app and was due to sending `onError` to a disposed emitter.